### DISCUSSION TITLE
Jinja2 templating

### DISCUSSION THREAD

**Discussion by jpdrsn** [pypyr/pypyr#324] [Q&A] [answered]
I was searching for how to combine Jinja2 and pypyr and got a result that said there was a jinja2 step function in pypyr - like the below. 

Is that a feature that used to be available? 

`
steps:
  - name: jinja2-template
    jinja2:
      template: |
        {% for item in items %}
        {{ item }}
        {% endfor %}
      context:
        items:
          - item1
          - item2
  `

**Comment by yaythomas** [comment, marked as answer]
Hi @jpdrsn,

There isn't a built-in a step for jinja, and there never was.

You can however make your own one, it's easy, promise!

If you have a custom .py file in your pipeline directory in a `steps` directory, like `./steps/jinja.py`:
```python
import logging
from pathlib import Path

from pypyr.context import Context
from pypyr.utils.asserts import assert_key_has_value

from jinja2 import Environment, FileSystemLoader

# getLogger will grab the parent logger context, so your loglevel and
# formatting automatically will inherit correctly from the pypyr core.
logger = logging.getLogger(__name__)


def run_step(context: Context) -> None:
    # friendly error msg if pipeline tries to use this without jinja input key
    context.assert_key_has_value(key='jinja', caller=__name__)

    # lets you use pypyr {format} expressions on inputs: dir, template, output
    step_input = context.get_formatted('jinja')

    # friendly err msg if the required child-keys of jinja aren't there, i.e:
    # jinja.template, jinja.output
    assert_key_has_value(obj=step_input,
                         key='template',
                         caller=__name__,
                         parent='jinja')

    assert_key_has_value(obj=step_input,
                         key='output',
                         caller=__name__,
                         parent='jinja')

    # defaults to current directory if not set
    templates_dir = step_input.get('dir', Path.cwd())
    template_file = step_input['template']
    outfile = step_input['output']

    # Use jinja api to generate output file and write to output file.
    file_loader = FileSystemLoader(templates_dir)
    env = Environment(loader=file_loader)
    template = env.get_template(template_file)
    output = template.render(**context)
   
    with open(outfile, 'w', encoding='utf-8') as file:
        file.write(output)

```

You can then have `./my-pipeline.yaml`:
```yaml
steps:
  - name: pypyr.steps.set
    in:
      set:
        fruit: apple
        vegetable: carrot
        arb_input: my-jinja-template.txt

  - name: steps.jinja
    comment: calling our custom jinja step!
    in:
      jinja:
        template: '{arb_input}'
        output: my-output-file.txt
```

Notice you can use the usual pypyr {formatting expressions} on all the inputs to the custom jinja step - so here the `jinja.template` input comes from the context key `arb_input` - no real reason other than just to show that you can if you wanted to. You could just use a plain string value here with the filename directly if you preferred.

The way I did it here is that jinja will use the pypyr context dictionary to find the variables it substitutes.

In other words, imagine `./my-jinja-template.txt` looks something like this:
```
This is my template. The fruit is {{ fruit }} and the vegetable is {{ vegetable }}.
```

**Reply by jpdrsn** [reply]
Thank you very much for that detailed example! Really good stuff.

**Reply by jpdrsn** [reply]
I had to make a minor adjustment to get it working on my system.
```
template = env.get_template(template_file)
output = template.render(**context)
```

**Reply by yaythomas** [reply]
whoops, sorry I missed that, fixed it up in the original sample also!
